nab/detectors/htmcore/htmcore_detector_pandaVis.py

```diff
@@ -0,0 +1,370 @@
+
+      # 3. Temporal Memory
+      # Execute Temporal Memory algorithm over active mini-columns.
+      #self.tm.compute(activeColumns, learn=True)
+
+      self.tm.activateCells(activeColumns, learn=True)
+      # activateDendrites calculates active segments
+      self.tm.activateDendrites(learn=True)
+      # predictive cells are calculated directly from active segments
+      predictiveCells = self.tm.getPredictiveCells()
+
+      self.tm_info.addData( self.tm.getActiveCells().flatten() )
+
+      self.PandaUpdateData(ts, val, encoding, activeColumns, predictiveCells)
+
+      # 4.1 (optional) Predictor #TODO optional
+      #TODO optional: also return an error metric on predictions (RMSE, R2,...)
+
+      # 4.2 Anomaly 
+      # handle spatial, contextual (raw, likelihood) anomalies
+      # -Spatial
+      spatialAnomaly = 0.0 #TODO optional: make this computed in SP (and later improve)
+      if self.useSpatialAnomaly:
+        # Update min/max values and check if there is a spatial anomaly
+        if self.minVal != self.maxVal:
+          tolerance = (self.maxVal - self.minVal) * SPATIAL_TOLERANCE
+          maxExpected = self.maxVal + tolerance
+          minExpected = self.minVal - tolerance
+          if val > maxExpected or val < minExpected:
+            spatialAnomaly = 1.0
+        if self.maxVal is None or val > self.maxVal:
+          self.maxVal = val
+        if self.minVal is None or val < self.minVal:
+          self.minVal = val
+
+      # -temporal (raw)
+      raw = self.tm.anomaly
+      temporalAnomaly = raw
+
+      if self.useLikelihood:
+        # Compute log(anomaly likelihood)
+        like = self.anomalyLikelihood.anomalyProbability(val, raw, ts)
+        logScore = self.anomalyLikelihood.computeLogLikelihood(like)
+        temporalAnomaly = logScore #TODO optional: TM to provide anomaly {none, raw, likelihood}, compare correctness with the py anomaly_likelihood
+
+      anomalyScore = max(spatialAnomaly, temporalAnomaly) # this is the "main" anomaly, compared in NAB
+
+      # 5. print stats
+      if self.verbose and self.iteration_ % 1000 == 0:
+          # print(self.enc_info)
+          # print(self.sp_info)
+          # print(self.tm_info)
+          pass
+
+      if PANDA_VIS_ENABLED:
+          pandaServer.BlockExecution()
+
+
+      return (anomalyScore, raw)
+
+  def BuildPandaSystem(self,modelParams):
+      global serverData
+      serverData = ServerData()
+      serverData.HTMObjects["HTM1"] = dataHTMObject()
+      serverData.HTMObjects["HTM1"].inputs["Input"] = dataInput()
+
+      serverData.HTMObjects["HTM1"].layers["Layer1"] = dataLayer(
+          modelParams["sp"]["columnCount"],
+          modelParams["tm"]["cellsPerColumn"],
+      )
+      serverData.HTMObjects["HTM1"].layers["Layer1"].proximalInputs = ["Input"]
+      serverData.HTMObjects["HTM1"].layers["Layer1"].distalInputs = ["Layer1"]
+
+
+  def PandaUpdateData(self, timestamp, value, encoding, activeColumns, predictiveCells):
```

Comment:
I'm thinking if merging this visualization code into the htmcore detector is a good idea? 
- the detector would be slightly more complex for the visualizations, but the extra code is not intrusive and clearly separated by `if PANDA_VIS_ENABLED` 
- we're still making some changes (small to detector itself, more to params), so I'd like to avoid mess in the duplicate code. 

What do you think, @Zbysekz ? Is it reasonable, or would it make the detector too complicated?

Reply:
>... to separate python file. That way only statements with PANDA_VIS_ENABLED will remain..
Do you think that then i should remove the htmcore_detector_pandaVis.py and make this into htm_core_detector.py ?

yes, this is what I'd like to see. If we can install htmpandavis core (the comm part) reasonably, the "intruding" part is not that complicated and clearly separated. So I think I'd be the best to have it merged in the detector.

Reply:
> That way only statements with PANDA_VIS_ENABLED will remain.. 

Apropos, for a later discussion, in htmcore, the Connections (so SP, TM) do support a "subscriber interface". Could we adapt that, and htmpandavis to visualize a htm system seamlessly? But first let's do the PR here to allow visualization in NAB for the detector.

Reply:
@Zbysekz do you want, will you have time to work on this? Otherwise I'd like to merge this as-is to get the "insight support" for NAB.

Reply:
yes, please give me few days, i will finish this. The installation of pandaComm folder works already 👍

Reply:
You rock! Awesome

Reply:
Ok ready to be merged, see other comment for resolution.